### PR TITLE
laser overclock for adv assline

### PR DESCRIPTION
**Pull request by Glease**
This enables adv assline to do "laser overclock" when TT energy hatch is used. It can overclock beyond voltage tier, but will also use more energy. (Still cannot do more advanced recipes). Every laser overclock will add 0.3 to power exponent.

e.g. Assume a recipe EU/t of 6000, recipe time of 2 minutes and 10 parallel. Uses 1 IV laser target hatch only. First column specify the max amp of that laser hatch.
```
 Laser  Hatch max  Peak EUt      Amp Recipe Overclock Imperfect Relative Parallel
  Size     Income     Usage     Used   time     times Overclock    Usage   Slices
     8      65530     60000      7.3   2400         0     60000   100.0%       10
    16     131070     60000      7.3   2400         0     60000   100.0%       10
    32     262140    258000     31.5   1200         1    240000   107.5%       10
    64     524280    258000     31.5   1200         1    240000   107.5%       10
   128    1048570    258000     31.5   1200         1    240000   107.5%       10
   256    2097150   1186800    144.9    600         2    960000   123.6%       10
   512    4194300   1186800    144.9    600         2    960000   123.6%       10
  1024    8388600   5815320    709.9    300         3   3840000   151.4%       10
  2048   16777210   5815320    709.9    300         3   3840000   151.4%       10
  4096   33554430  30239660   3691.4    150         4  15360000   196.9%       10
  8192   67108860  30239660   3691.4    150         4  15360000   196.9%       10
 16384  134217720  30239660   3691.4    150         4  15360000   196.9%       10
 32768  268435450 166318160  20302.5     75         5  61440000   270.7%       10
 65536  536870910 166318160  20302.5     75         5  61440000   270.7%       10
131072 1073741820 964645520 117754.6     37         6 245760000   392.5%       10
262144 2147483640 964645520 117754.6     37         6 245760000   392.5%       10
```
Imperfect overclock column is the EU usage under normal imperfect overclock and the same target recipe time. 
Third column is labelled as **peak** usage because the machine will have less than 10 slices active at the very beginning of the craft. If your batch is small enough you might never see a EU/t this high, but the machine will overclock with the assumption that there will be a moment that all 10 slices are active.

For easy copy pasting, it is not formatted using github table syntax.

Am I too generous, or is this too harsh? Please comment!

### CODE REVIEW

**Comment by BlueWeabo** [comment]
I think it will be a good idea to lower how much more energy is being consumed. Not to push anything, but currently using anything above a 64 A energy hatch will be practically worthless imo.

**Comment by BlueWeabo** [comment]
In short. The player will be more likely to spam the multi instead of Overclocking it at a certain point. At IV the energy increases aren't that big, but at UMV and beyond they will be quite a lot more power if one tries to OC them, but those recipes have such a high time it might be worth it(prolly not cus of the extra power needed)

**Comment by minecraft7771** [comment]
Couldn't you cheese progression using this? Lasers are available in ZPM using the GT++ assembler and low power lasers are available in I think LuV even.

**Comment by BlueWeabo** [comment]
No it will not allow for cheesing. This only allows players to overclock their recipes above their tier at a cost of extra power.

**Comment by BlueWeabo** [comment]
Yes the voltage of the hatch is used, while the extra amps can allow for extra overclock

**Comment by Glease** [comment]
I forgot to factor in the parallel in this simulation. I have updated the table.

**Comment by BlueWeabo** [comment]
If I have understood correctly only the entire assline consumed the power and not each slice right?

**Comment by Glease** [comment]
I'm not sure what you are referring to exactly, but the total EU/t at any given time is the *recipe EU/t after all overclock* times *number of active slices* (not including STUCK slices)

**Comment by BlueWeabo** [comment]
So each slice is drawing the recipe EU/t after overclocks. Also did you make the laser OC start later? Aka now it seems to start from 8A instead of 1A

**Comment by Glease** [comment]
First Parallel, then normal OC, last laser OC. They are calculated in the order of energy efficiency, as parallel is perfect OC (in a sense), normal OC loss 50% every OC, laser OC loss even more every OC)

**Comment by BlueWeabo** [comment]
How would the graph change if it was 0.25 power exponent instead of a 0.3?

**Comment by Glease** [comment]
This is what it look like at 0.25
```
 Laser  Hatch max  Peak EUt      Amp Recipe Overclock Imperfect Relative Parallel
  Size     Income     Usage     Used   time     times Overclock    Usage   Slices
     8      65530     60000      7.3   2400         0     60000   100.0%       10
    16     131070     60000      7.3   2400         0     60000   100.0%       10
    32     262140    255000     31.1   1200         1    240000   106.3%       10
    64     524280    255000     31.1   1200         1    240000   106.3%       10
   128    1048570    255000     31.1   1200         1    240000   106.3%       10
   256    2097150   1147500    140.1    600         2    960000   119.5%       10
   512    4194300   1147500    140.1    600         2    960000   119.5%       10
  1024    8388600   5450620    665.4    300         3   3840000   141.9%       10
  2048   16777210   5450620    665.4    300         3   3840000   141.9%       10
  4096   33554430  27253100   3326.8    150         4  15360000   177.4%       10
  8192   67108860  27253100   3326.8    150         4  15360000   177.4%       10
 16384  134217720  27253100   3326.8    150         4  15360000   177.4%       10
 32768  268435450 143078780  17465.7     75         5  61440000   232.9%       10
 65536  536870910 143078780  17465.7     75         5  61440000   232.9%       10
131072 1073741820 786933280  96061.2     37         6 245760000   320.2%       10
262144 2147483640 786933280  96061.2     37         6 245760000   320.2%       10
```

**Comment by BlueWeabo** [comment]
can you fix the bug of the advanced assline not consuming inputs from ME stocking input busses?

**Comment by Glease** [comment]
fixed

File: src/main/resources/assets/ggfab/lang/en_US.lang [hunk 1]
```diff
@@ -4,8 +4,13 @@ ggfab.info.advassline.2=It supports item pipelining. That is, it will mimic a re
...
-ggfab.info.advassline.6=The EU/t cost of this machine is number of slices active multiplied by the original recipe EU/t. §4STUCK§r slices do not consume power.
-ggfab.waila.advassline.slice=Slice %s: %s s/%s s
-ggfab.waila.advassline.slice.small=Slice %s: %s ticks remaining
-ggfab.waila.advassline.slice.idle=Slice %s: Idle
-ggfab.waila.advassline.slice.stuck=Slice %s: §4STUCK
\ No newline at end of file
+ggfab.info.advassline.6=The EU/t cost of this machine is number of slices active multiplied by the original recipe EU/t. §4STUCK§r slices do not consume power. It will use the worst energy supplying hatch's input voltage for recipe tier calculation and normal imperfect overclock.
+ggfab.info.advassline.7=With exotic energy hatches, it can overclock beyond usual voltage tier, but will consume even more power than usual imperfect overclock. Every §2laser overclock§r will add 0.5 to power exponent.
```
Comment (BlueWeabo) [inline src/main/resources/assets/ggfab/lang/en_US.lang:8]: Isn't the power exponent 0.3?

File: src/main/java/net/glease/ggfab/util/OverclockHelper.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    public static OverclockOutput normalOverclock(long recipeEUt, int duration, long inputVoltage, boolean perfectOC) {
+        if (recipeEUt > inputVoltage) return null;
+        int recipeTier = Math.max(1, GT_Utility.getTier(recipeEUt)); // ULV no overclock
+        int machineTier = GT_Utility.getTier(inputVoltage);
+        int shift = perfectOC ? 2 : 1;
+        while (recipeTier < machineTier && duration > 1) {
+            duration >>= shift;
+            recipeEUt <<= 2;
+            recipeTier++;
+        }
```
Comment (BlueWeabo) [inline src/main/java/net/glease/ggfab/util/OverclockHelper.java:15]: As I general rule I think you should try and make sure you aren't accidentally Overclocking one too many times. Has happened to me previously where I have 1 UMV energy hatch, but it tries to OC to UXV and fails.

Comment (Glease) [reply src/main/java/net/glease/ggfab/util/OverclockHelper.java:15]: There are unit tests for this overclock helpers. If you find a situation that it doesn't meet the expectation, please state the case clearly so I can add it to the tests and see what went wrong.

File: src/main/resources/assets/ggfab/lang/en_US.lang (hunk 1 above)
Comment (Glease) [reply src/main/resources/assets/ggfab/lang/en_US.lang]: Fixed.

**Comment by Glease** [comment]
Screw the balance. I'll just make the penalty factor a config variable and let whoever has the power to tweak it.

File: src/main/java/net/glease/ggfab/util/OverclockHelper.java (hunk 2 above)
Comment (BlueWeabo) [reply src/main/java/net/glease/ggfab/util/OverclockHelper.java:15]: Alright if anything like that happens I will send a message.

File: src/main/java/net/glease/ggfab/ConfigurationHandler.java [hunk 3]
```diff
@@ -30,7 +30,7 @@ private void setLanguageKeys() {
     }
 
     private void loadConfig() {
-        laserOCPenaltyFactor = config.getFloat("advasslinePenaltyFactor", "common.balancing", 0.25f, 0f, 10f, "Laser overclock penalty factor. This will incredibly change the game balance. Even a small step from 0.3 to 0.3 can have very significant impact. Tweak with caution!");
+        laserOCPenaltyFactor = config.getFloat("advasslinePenaltyFactor", "common.balancing", 0.3f, 0f, 10f, "Laser overclock penalty factor. This will incredibly change the game balance. Even a small step from 0.3 to 0.3 can have very significant impact. Tweak with caution!");
```
Comment (BlueWeabo) [inline src/main/java/net/glease/ggfab/ConfigurationHandler.java:33]: What do you mean a change from 0.3 to 0.3 the same number can have a significant impact

**Review by BlueWeabo — APPROVED** [review]

**Pull request merged**